torch_geometric/nn/conv/point_transformer_conv.py

```diff
@@ -0,0 +1,144 @@
+from typing import Union
+from torch_geometric.typing import PairOptTensor, PairTensor, Adj
+
+from torch import Tensor
+from torch_sparse import SparseTensor, set_diag
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.utils import remove_self_loops, add_self_loops
+
+from torch.nn import Sequential, ReLU, Linear
+from torch_geometric.utils import softmax
+
+from ..inits import reset
+
+
+class PointTransformerConv(MessagePassing):
+    r"""The PointTransformer layer from the `"Point Transformer"
+    <https://arxiv.org/abs/2012.09164>` paper
+
+    .. math::
+        \mathbf{x}^{\prime}_i =  \sum_{j \in
+        \mathcal{N}(i) \cup \{ i \}} \rho \left( \gamma \left( \phi
+        \left( \mathbf{x}_i \right) - \psi \left( \mathbf{x}_j \right) \right)
+        + \delta  \right) \odot \left( \alpha \left(  \mathbf{x}_j \right)
+        + \delta \right) ,
+
+    where :math:`\gamma` and :math:`\delta = \theta \left( \mathbf{p}_j -
+    \mathbf{p}_i \right)` denote neural networks, *.i.e.* MLPs, that
+    respectively map transformed node features and relative spatial
+    coordinates. :math:`\mathbf{P} \in \mathbb{R}^{N \times D}`
+    defines the position of each point.
+
+    Args:
+        in_channels (int): Size of each input sample features
+        out_channels (int) : size of each output sample features
+        local_nn : (torch.nn.Module, optional): A neural network
+            :math:`\delta = \theta \left( \mathbf{p}_j -
+            \mathbf{p}_i \right)` which maps relative spatial coordinates
+            :obj:`pos_j - pos_i` of shape
+            :obj:`[-1, 3]` to shape
+            :obj:`[-1, out_channels]`, *e.g.*, defined by
+            :class:`torch.nn.Sequential`. In case :obj:`None` is given, is
+            initialized with :obj:`Sequential(\n'
+            '    (0): Linear(in_features=3, out_features=64, bias=True)\n'
+            '    (1): ReLU()\n'
+            '    (2): Linear(in_features=64, out_features=32, bias=True)\n'
+            '  )\n'` (default: :obj:`None`)
+        global_nn : (torch.nn.Module, optional): A neural network
+            :math:`\gamma` which maps transformed node features of shape
+            :obj:`[-1, out_channels]` to shape :obj:`[-1, out_channels]`,
+            *e.g.*, defined by :class:`torch.nn.Sequential`. In case
+            :obj:`None` is given, is initialized with :obj:`Sequential(\n'
+            '  (0): Linear(in_features=32, out_features=64, bias=True)\n'
+            '  (1): ReLU()\n'
+            '  (2): Linear(in_features=64, out_features=32, bias=True)\n'
+            ')'` (default: :obj:`None`)
```

Comment:
@rusty1s , I suggested we set the default values of `local_nn` and `global_nn` to the network architecture in the paper, so users have access to that by default. Do you think thats the right way to go here, or should `local_nn` and `global_nn` be non-optional.

Reply:
@wsad1 @rusty1s 
The main constraint is that local_nn maps from dim `[-1,3]` to `[-1, out_channels]` so it **must** be initialized. 
It can't be put as default argument as it depends on another argument. I had a look at Linear from torch_geometric which allows lazy initialization, however it can only do so with input channel.

I added a commit "no default nns" to see how it could be done if we leave local_nn and global_nn to the user. While it makes the code shorter so simpler, it leaves out the default network architectures of the paper. PointConv does it that way but it has an example for himself in this repo. But to me a user expects the default values to be at least acceptable (which is not the case if we leave the nns to the user charge), and should not have to look into the paper to define the corresponding `local_nn ` and `global_nn `.

With accessibility in mind, I think the best way to go is the first way I did it and as you say, with the networks being non optional. `(in_channels, out_channels, mlp_pos, attention_dim)`  would be the arguments and `local_nn` and `global_nn` would be defined in the function. An experienced user will still have the possibility to change it manually by redefining the attribute. Although it is ugly, I think it will only concern a minority of nitpicking users while making it more accessible to the majority.

What are your thoughts on it ?